### PR TITLE
Add a headless emu server to allow for automated testing of the emulator

### PR DESCRIPTION
**Pull request by solsjo**
- brickEmu has high coupling to brickOs-bibo, this is a starting point
  for making sure that the two repositories are compatible

- A scriptable emu server interface allows for simulating stimuli and
  response in a test driven manner

### CODE REVIEW

**Comment by mesheets** [comment]
Appreciate the interest and pull request, @solsjo!  A test harness for BrickEmu itself would definitely be useful.

Based on the default port, is this intended to be compatible with software such as [ir-server](https://github.com/BrickBot/brickOS-bibo/blob/9ffac7edd506db38724263a8c2455df5dbfb84bd/README#L36) (c.f. [note](https://github.com/BrickBot/brickOS-bibo/blob/9ffac7edd506db38724263a8c2455df5dbfb84bd/README#L48)) / [lnpd](https://github.com/BrickBot/lnpd) / [lnphost](https://github.com/BrickBot/lnphost) and [others](https://github.com/search?q=org%3ABrickBot+LNP), which basically virtualize the IR communication of the RCX?  The default IR virtualization port has been the "magic" LEGOS port of [50637](https://github.com/BrickBot/brickOS-bibo/blob/9ffac7edd506db38724263a8c2455df5dbfb84bd/util/host/ir-server.c#L70).



For any who might not be familiar with the IR virtualization, I will try to include a brief note here—
While there is a closer coupling with BrickOS-bibo for some of the "shortcut" or quick access features of BrickEmu, the core emulator functionality of BrickEmu _should_ be compatible with any non-BrickOS-bibo firmware if using the IR virtualization capabilities.  While ir-server is currently included as part of the BrickOS-bibo build, it is similar in nature to the firmware download tooling also built with BrickOS-bibo, in that it's usage is more general.

Some software includes built-in support for IR virtualization by allowing a network address and port to be specified, but interfacing with other software that does not directly support the IR virtualization is possible by using the [uplink/downlink](https://github.com/BrickBot/brickOS-bibo/blob/9ffac7edd506db38724263a8c2455df5dbfb84bd/util/host/ir-server.c#L116) functionality of ir-server to interface with a serial port (which can either be a physical serial port connected to an actual IR tower or a virtual serial port [such as com-0-com on Windows] created for interfacing with software expecting a serial port connection).  The [hubub](https://github.com/BrickBot/hubub) project might also be of use in certain situations.

I'm sure there are things that could be documented better or more clearly, too.  Perhaps it would be appropriate to move some of these more general tools and utilities (e.g. firmware/program download and ir-server) from the BrickOS-bibo repository to the BrickEmu repository?

**Comment by solsjo** [comment]
Hi!

- **Based on the default port, is this intended to be compatible with software such as ir-server**

Yes it is, or rather, I intentionally set the scope to leave out the implementation of the IR communication with the Emulator. Since there existed so many tools. That way the ir-server and nqc or dll or whatever tool chosen is the responsibility of the user/usecase at least for now. :) I unfortunately named it  "EMU_SERVER_TCP_PORT" in emu_server.py it should be something similar to IR_TCP_PORT.

- **I'm sure there are things that could be documented better or more clearly, too. Perhaps it would be appropriate to move some of these more general tools and utilities (e.g. firmware/program download and ir-server) from the BrickOS-bibo repository to the BrickEmu repository?**

For sure, maybe even consolidating some of the tools.

As a follow up to this PR it would be nice to add a control command/interface controlling the stepping of the clock/tick/instructions from an external part, e.g. from the emu_server.
That way one could simulate e.g. the linetracker robot and the line to follow visually in e.g. in a python GUI. or other framework with greater precision. So that you can test sensor/motor stimuli in simulation.

With that said, if I rename the port, is this PR a candidate for merge?

**Comment by mesheets** [comment]
To be fair, it is an emulated IR port.  :-)

My availability for the BrickBot project has tended to be in spurts, and there are actually additional RCX-related projects that I have yet to work through to get them posted.  Would you be interested in being added to this project?

**Comment by mesheets** [comment]
@solsjo, as of the last prior commit to this project about 6 months ago, the CI action [ran successfully](https://github.com/BrickBot/brickEmu/actions/runs/10615522954) (including pyTest).

Before making any new commits, I manually ran the CI action to verify that everything in the CI action still worked with the latest version of the GitHub runners.  Unfortunately, pyTest is [now hanging this CI action against the exact same code](https://github.com/BrickBot/brickEmu/actions/runs/13252268593/job/36992544148), apparently due to the following:
`*** buffer overflow detected ***: terminated`

Despite that `terminated` message, though, pyTest just hangs.

Might you have any idea what might have happened?  Was something perhaps deprecated that pyTest was using?

**Comment by solsjo** [comment]
Without knowing, a set of guesses would be that:
TL;DR;
python version has been bumped. And the used function deprecated, such that the server never goes active. 
similar problem here, [pypa](https://github.com/pypa/setuptools/issues/2643)
I think the solution then, is, change
emu_serv.setDaemon(True)
to
emu_serv.daemon = True



looking at the actions and and the setup. I realize two things.
1) We don’t have control over ubuntu version. Meaning github will determine it for us. And we won’t have obvious versioned control regarding which OS we can run on.
2) python version is what is latest on that ubuntu version.
3) requesting pip install without specifying version will get us the latest pytest everytime we run the action. Meaning, pytest might have signaled deprication notice of set_dameon() for quite a while and we notice the side effect just now as it does not work anymore.

**Comment by mesheets** [comment]
I appreciate the follow up.  I've made a few updates (including to the [GitHub workflow](https://github.com/BrickBot/brickEmu/blob/master/.github/workflows/brickEmu_CI.yml)) adding matrix values for the Ubuntu version.  The test runs fine on Ubuntu 20.04 and 22.04 but not on 24.04 / latest ([link to a GitHub Actions run](https://github.com/BrickBot/brickEmu/actions/runs/13277745032)), yet a look at the logs appears to show the same Python and PIP versions being installed on each OS:
* Python version was [set in the matrix as 3.12](https://github.com/BrickBot/brickEmu/blob/50189cf304805b3671064a82a7765b2d112c20f7/.github/workflows/brickEmu_CI.yml#L18), so I did not change this.  The specific version[ reported on the Ubuntu 20.04, 22.04 and 24.04 / Latest runners was 3.12.8](https://github.com/BrickBot/brickEmu/actions/runs/13277745032/job/37070309658#step:8:10).
* pytest version reported on both the passing and failing runners was 8.3.4, which appears to be the [current latest pytest version](https://github.com/pytest-dev/pytest/releases).
* Changed how [daemon is set](https://github.com/BrickBot/brickEmu/blob/50189cf304805b3671064a82a7765b2d112c20f7/tests/conftest.py#L35), as advised

PIP version install comparison:
* PIP on 22.04: `Successfully installed flake8-7.1.1 iniconfig-2.0.0 mccabe-0.7.0 packaging-24.2 pluggy-1.5.0 pycodestyle-2.12.1 pyflakes-3.2.0 pytest-8.3.4`
* PIP on 24.04: `Successfully installed flake8-7.1.1 iniconfig-2.0.0 mccabe-0.7.0 packaging-24.2 pluggy-1.5.0 pycodestyle-2.12.1 pyflakes-3.2.0 pytest-8.3.4`

**Comment by solsjo** [comment]
Aha! Great investigation!

**Comment by solsjo** [comment]
Aha! Great investigation!

**Comment by solsjo** [comment]
A bit out of ideas, but great that you pinpointed it to 24.04.
Will try to look at it a bit later

**Comment by solsjo** [comment]
```

pytest.emu = subprocess.Popen(
            [emu_bin, "-serverport", f"{pytest.emu_obj.port}"], stdout=subprocess.PIPE
        )

        while not pytest.emu_obj.emulator_status_up():
            time.sleep(0.2)
            continue

```

That while loop seems like a likely candidate for a hang, perhaps we ought to have a timeout condition for it, to not loop indefinitely.

If it is the case that it waits at this loop, then it seems reasonable that either emu_obj or emu server runs into problems.

If possible then, it would be nice to see how, each command executes on 24.04.

We should be able to trigger/log it via the test if we add a condition to exit the loop after a few seconds and then inspecting which binary that got a buffer overflow

**Comment by solsjo** [comment]
E.g.

```
time_sleept = 0
timeout = 5
while not pytest.emu_obj.emulator_status_up() and time_sleept < timeout:
            time.sleep(0.2)
            time_sleept += 0.2

            continue

if time_sleept > timeout:
    emu_server.exit()
    raise SomeError([emu_obj_process.stdout, emu_server.log_if_it_exists]) # something like that

```

NOTE: pseudo code

**Comment by mesheets** [comment]
One item that is concretely reporting as different is GCC.  It perhaps even could be a combination of the Ubuntu version and the GCC version, as in the past Ubuntu has built their GCC packages with certain stricter flags set (FORTIFY_SOURCE being one, as I recall).

| Ubuntu Version | GCC Version |
|---------------|-------------|
| 20.04 | 9.4.0 |
| 22.04 | 11.4.0 |
| 24.04 (latest) | 13.3.0 |

Apparently, among other software, [a similar "buffer overflow detected" issue was encountered with 7-Zip when migrating from Ubuntu 22.04 to 24.04](https://sourceforge.net/p/sevenzip/support-requests/581/).  Ubuntu 20.04 was [deprecated](https://github.com/actions/runner-images/issues/11101) as a GitHub Actions runner at the beginning of the month, so that one won't be around much longer.

Unfortunately, this BrickEmu code triggers a plethora of compiler warnings.  I've seen in other projects where GCC would output a warning if—based on given substitutions or expansions—the maximum possible string length exceeded the size of the buffer, but offhand I'm not seeing a warning like that (unless perhaps I'm missing it amid all the other warning messages).

Even with compiler flags `-g -fsanitize=address`, the `*** buffer overflow detected ***: terminated` message doesn't provide much insight.

**Comment by solsjo** [comment]
Ack!
Great!
Yes, I came across some bug-reports as well, but no obvious debugging clues. Applying stricter flags sounds like a good plan.
Maybe running it with valgrind would provide more information?
Otherwise kicking up a debugger?

**Comment by mesheets** [comment]
Found a [Valgrind action](https://github.com/marketplace/actions/valgrind-checker) and added it.  Unfortunately, it doesn't appear to have flagged anything, but perhaps it takes more test-like activity to trigger the overflow….  :-/

**Comment by solsjo** [comment]
Ack, we could try running it in the test script e.g. 
```
subprocess.run(["valgrind", "-some-flag", emu_obj])
```

So that, as you say, we run it in the same env as where the error is triggered. At this point, I do not have an environment to test this, it's somewhat annoying to have to push commit after commit, to try to pinpoint the error. The feedback loop is pretty slow.

**Comment by solsjo** [comment]
Perhaps also compile with debug symbols, e.g. -ggdb3 unless we already do

**Comment by solsjo** [comment]
I also assume that we don't know which binary that triggers the overflow?

```
    firmware = "../brickos_bibo/kernel/bibo.coff"
    ir_server_bin = "./ir-server"
    emu_bin = "./emu"
    ir_client_bin = "./../brickos_bibo/util/dll"
    program = "./../brickos_bibo/demo/c/linetrack.lx"
```

When running valgrind, one could do something similar to this:

valgrind --tool=exp-sgcheck <binary>
e.g.
```

pytest.emu = subprocess.Popen(
            ["valgrind", "--tool=exp-sgcheck", emu_bin, "-serverport", f"{pytest.emu_obj.port}"], stdout=subprocess.PIPE
        )

```

**Comment by mesheets** [comment]
> At this point, I do not have an environment to test this, it's somewhat annoying to have to push commit after commit, to try to pinpoint the error. The feedback loop is pretty slow.

I was having pretty much the same thoughts myself, lol.


> Perhaps also compile with debug symbols, e.g. -ggdb3 unless we already do

[Current CFLAGS](https://github.com/BrickBot/brickEmu/blob/2abac7b88dd4581d6132c2249ed8cc718db6793e/Makefile#L2C8-L2C64) are `-O2 -pg -g -fsanitize=address -Wall -Wmissing-prototypes`.  The `-fsanitize=address` is a more recent addition.  Open to any suggestions or changes here.

> I also assume that we don't know which binary that triggers the overflow?

Unless someone else has been able to identify something, I'm not sure yet.  Based on how the failures seemed to come about as a result of a newer Ubuntu version, it would seem likely to me to be in binaries generated by the native GCC compiler (e.g. emu, ir-server, dll, firmdl) and not the H8/300 cross-compiler (e.g. coff, srec, and *.lx).

Support for the H8/300 was dropped from GCC a long time ago, and the H8/300 cross-compiler Ubuntu packages for binutils and gcc are still based on [binutils 2.16.1](https://launchpad.net/ubuntu/+source/binutils-h8300-hms) and [gcc 3.4.6](https://launchpad.net/ubuntu/+source/gcc-h8300-hms).

**Comment by mesheets** [comment]
…interestingly, the [Valgrind run failed](https://github.com/BrickBot/brickEmu/blob/8492d5ce7c7d5e98f035df099f38bf6d5ff79ec7/Makefile#L7) when I changed the CI build make command to `make SOUND=mute` (c.f. [Makefile](https://github.com/BrickBot/brickEmu/blob/8492d5ce7c7d5e98f035df099f38bf6d5ff79ec7/Makefile#L7)).  Reverted that, and the Valgrind checks passed again.

My thought was that if things worked without sound, perhaps the problem is somewhere within the sound support, but apparently something more is going on.

**Comment by solsjo** [comment]
Ack, yes, some kind of bisecting might be good approach.
I also notice that the python code only connects to the stdout, but not the stderr. I wonder if there is information there that we miss out on. I'm also wondering if clang might provide better error reporting with the similar sanitize feature.

But as you said, it would be nice to be able to replicate locally, perhaps with a docker container.

**Comment by solsjo** [comment]
@mesheets , just wanted to let you know that you can set up a github codespace online, e.g. the blank template and then once up and running change the default image from microsoft's default 20.04 to the 24.04 image,
.devcontainer.json
`{"image":"ubuntu-24.04"}`

from saving the file, github will ask if you want to rebuild.
after which:

`gcc --version
gcc (Alpine 13.2.1_git20240309) 13.2.1 20240309`

**Comment by solsjo** [comment]
or `{"image":"mcr.microsoft.com/devcontainers/cpp:dev-ubuntu24.04"}`

**Comment by solsjo** [comment]
after going thru hurdles with installing all deps,

reproduced:

```
sudo apt-get update
sudo apt-get --assume-yes install binutils-h8300-hms gcc-h8300-hms zlib1g-dev libsdl-sound1.2-dev
make
sudo apt-get install pip
sudo apt install python3-pytest
python3 -m pytest ./tests
*** buffer overflow detected ***: terminated
```

**Comment by solsjo** [comment]
I'm pretty sure emu is the application that triggers the buffer overflow.

**Comment by solsjo** [comment]
```
./emu -serverport 35143
serverport=35143
*** buffer overflow detected ***: terminated
Aborted (core dumped)
```

```
valgrind --leak-check=full --show-leak-kinds=all ./emu -serverport 35143
==6591== Memcheck, a memory error detector
==6591== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==6591== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==6591== Command: ./emu -serverport 35143
==6591== 
==6591==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
==6591== 
==6591== HEAP SUMMARY:
==6591==     in use at exit: 373,992 bytes in 1 blocks
==6591==   total heap usage: 1 allocs, 0 frees, 373,992 bytes allocated
==6591== 
==6591== 373,992 bytes in 1 blocks are still reachable in loss record 1 of 1
==6591==    at 0x484D953: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==6591==    by 0x518701F: monstartup (gmon.c:175)
==6591==    by 0x114A20: __gmon_start__ (in /workspaces/brickEmu/emu)
==6591==    by 0x4846015: ??? (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==6591==    by 0x40058C5: call_init (dl-init.c:37)
==6591==    by 0x40058C5: _dl_init (dl-init.c:88)
==6591==    by 0x401F59F: ??? (in /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2)
==6591==    by 0x2: ???
==6591==    by 0x1FFF000026: ???
==6591==    by 0x1FFF00002C: ???
==6591==    by 0x1FFF000038: ???
==6591== 
==6591== LEAK SUMMARY:
==6591==    definitely lost: 0 bytes in 0 blocks
==6591==    indirectly lost: 0 bytes in 0 blocks
==6591==      possibly lost: 0 bytes in 0 blocks
==6591==    still reachable: 373,992 bytes in 1 blocks
==6591==         suppressed: 0 bytes in 0 blocks
==6591== 
==6591== For lists of detected and suppressed errors, rerun with: -s
==6591== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by solsjo** [comment]
@mesheets 
Apparently when running via valgrind, we need to remove the `-fsantize=address` flag,
so then we get:
```
ERROR:
==4320== Memcheck, a memory error detector
==4320== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==4320== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==4320== Command: ./emu -serverport 49303
==4320== 
*** buffer overflow detected ***: terminated
==4320== 
==4320== Process terminating with default action of signal 6 (SIGABRT): dumping core
==4320==    at 0x4A25B2C: __pthread_kill_implementation (pthread_kill.c:44)
==4320==    by 0x4A25B2C: __pthread_kill_internal (pthread_kill.c:78)
==4320==    by 0x4A25B2C: pthread_kill@@GLIBC_2.34 (pthread_kill.c:89)
==4320==    by 0x49CC27D: raise (raise.c:26)
==4320==    by 0x49AF8FE: abort (abort.c:79)
==4320==    by 0x49B07B5: __libc_message_impl.cold (libc_fatal.c:134)
==4320==    by 0x4ABDC18: __fortify_fail (fortify_fail.c:24)
==4320==    by 0x4ABD5D3: __chk_fail (chk_fail.c:28)
==4320==    by 0x4ABDE64: __fread_chk (fread_chk.c:39)
==4320==    by 0x1168DE: fread (stdio2.h:215)
==4320==    by 0x1168DE: coff_symbols (coff.c:86)
==4320==    by 0x110FFF: read_rom (memory.c:193)
==4320==    by 0x111092: mem_init (memory.c:232)
==4320==    by 0x10BC14: main (main.c:83)
==4320== 
==4320== HEAP SUMMARY:
==4320==     in use at exit: 160,043 bytes in 5 blocks
==4320==   total heap usage: 5 allocs, 0 frees, 160,043 bytes allocated
==4320== 
==4320== 472 bytes in 1 blocks are still reachable in loss record 1 of 5
==4320==    at 0x4846828: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4320==    by 0x4A0CE7E: __fopen_internal (iofopen.c:65)
==4320==    by 0x4A0CE7E: fopen@@GLIBC_2.2.5 (iofopen.c:86)
==4320==    by 0x110F53: read_rom (memory.c:189)
==4320==    by 0x111092: mem_init (memory.c:232)
==4320==    by 0x10BC14: main (main.c:83)
==4320== 
==4320== 4,096 bytes in 1 blocks are still reachable in loss record 2 of 5
==4320==    at 0x4846828: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4320==    by 0x4A0C1B4: _IO_file_doallocate (filedoalloc.c:101)
==4320==    by 0x4A1C523: _IO_doallocbuf (genops.c:347)
==4320==    by 0x4A19F8F: _IO_file_overflow@@GLIBC_2.2.5 (fileops.c:745)
==4320==    by 0x4A1AAAE: _IO_new_file_xsputn (fileops.c:1244)
==4320==    by 0x4A1AAAE: _IO_file_xsputn@@GLIBC_2.2.5 (fileops.c:1197)
==4320==    by 0x49E7CC8: __printf_buffer_flush_to_file (printf_buffer_to_file.c:59)
==4320==    by 0x49E7CC8: __printf_buffer_to_file_done (printf_buffer_to_file.c:120)
==4320==    by 0x49F2742: __vfprintf_internal (vfprintf-internal.c:1545)
==4320==    by 0x4ABEA0C: __printf_chk (printf_chk.c:33)
==4320==    by 0x10BC0A: printf (stdio2.h:86)
==4320==    by 0x10BC0A: main (main.c:75)
==4320== 
==4320== 4,096 bytes in 1 blocks are still reachable in loss record 3 of 5
==4320==    at 0x4846828: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4320==    by 0x4A0C1B4: _IO_file_doallocate (filedoalloc.c:101)
==4320==    by 0x4A1C523: _IO_doallocbuf (genops.c:347)
==4320==    by 0x4A1AF4B: _IO_file_xsgetn (fileops.c:1288)
==4320==    by 0x4A0D49A: fread (iofread.c:38)
==4320==    by 0x1167D6: fread (stdio2.h:212)
==4320==    by 0x1167D6: coff_init (coff.c:64)
==4320==    by 0x110F65: read_rom (memory.c:191)
==4320==    by 0x111092: mem_init (memory.c:232)
==4320==    by 0x10BC14: main (main.c:83)
==4320== 
==4320== 4,355 bytes in 1 blocks are still reachable in loss record 4 of 5
==4320==    at 0x4846828: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4320==    by 0x1168B5: coff_symbols (coff.c:85)
==4320==    by 0x110FFF: read_rom (memory.c:193)
==4320==    by 0x111092: mem_init (memory.c:232)
==4320==    by 0x10BC14: main (main.c:83)
==4320== 
==4320== 147,024 bytes in 1 blocks are still reachable in loss record 5 of 5
==4320==    at 0x484D953: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4320==    by 0x4AB401F: monstartup (gmon.c:175)
==4320==    by 0x10BD70: __gmon_start__ (in /workspaces/brickEmu/emu)
==4320==    by 0x4846015: ??? (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==4320==    by 0x40058C5: call_init (dl-init.c:37)
==4320==    by 0x40058C5: _dl_init (dl-init.c:88)
==4320==    by 0x401F59F: ??? (in /usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2)
==4320==    by 0x2: ???
==4320==    by 0x1FFF000052: ???
==4320==    by 0x1FFF000058: ???
==4320==    by 0x1FFF000064: ???
==4320== 
==4320== LEAK SUMMARY:
==4320==    definitely lost: 0 bytes in 0 blocks
==4320==    indirectly lost: 0 bytes in 0 blocks
==4320==      possibly lost: 0 bytes in 0 blocks
==4320==    still reachable: 160,043 bytes in 5 blocks
==4320==         suppressed: 0 bytes in 0 blocks
==4320== 
==4320== For lists of detected and suppressed errors, rerun with: -s
==4320== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

were:
**by 0x1168DE: fread (stdio2.h:215)
by 0x1168DE: coff_symbols (coff.c:86)
by 0x110FFF: read_rom (memory.c:193)
by 0x111092: mem_init (memory.c:232)
by 0x10BC14: main (main.c:83)**
is suspect, however as a buffer overflow the origin might be elsewhere, but with a bit luck here.

it is worth looking into the other calls as well.
but putting a print/breakpoint close to line 86 in coff.c or 193 in memory.c

might be a good approach

**Comment by solsjo** [comment]
Adding some prints seems to reveal that malloc fails for:
```
    fseek(file, symptr + 18 * nsyms, SEEK_SET);
    ret = fread(&buf, 4, 1, file);
    fprintf("read_length: %d", ret);
    length = ntohl(*(uint32 *)buf);
    symtable = malloc(length);
    uint32 sym_ptr = ntohl(*(uint32_t *)symtable); 
    fprintf(stderr, "");
    fprintf(stderr, "\nlen %d\n", length);
    fprintf(stderr, "\nsymtable %d\n", sym_ptr);
```

```
FILE; ./rom.coff
read_length: 1
len 4355

symtable 0
*** buffer overflow detected ***: terminated
```

but malloc failing seems like a weird error which might point to the actual error existing elsewhere?

**Comment by solsjo** [comment]
having added some prints, it seems as symtable pointer is NULL, i.e. malloc in 
```
    coff.c:85
    

    length = ntohl(*(uint32 *)buf);
    symtable = malloc(length);
    fread(symtable+4, 1, length, file);
```

**Comment by mesheets** [comment]
This is some really good information you've found!  I have removed `-fsantize=address` from CFLAGS, and with that, the Valgrind GitHub action now appears to be failing outright on all Ubuntu versions (20.04, 22.04, and 24.04).

**Comment by solsjo** [comment]
@mesheets 
```
git diff coff.c
diff --git a/coff.c b/coff.c
index 74d1bdb..b6a02b3 100644
--- a/coff.c
+++ b/coff.c
@@ -82,7 +82,7 @@ void coff_symbols (FILE *file, int start)
     fseek(file, symptr + 18 * nsyms, SEEK_SET);
     fread(&buf, 4, 1, file);
     length = ntohl(*(uint32 *)buf);
-    symtable = malloc(length);
+    symtable = malloc(length+4);
     fread(symtable+4, 1, length, file); 
```
The code seemed to read data into symtable from at an offset  4, while the size of the data is equal to the size of the malloc 
call meaning we'll write outside of the allocated buffer?
This seems to resolve it for me, can you check try out a commit and see it resolves it in the pipeline.

**Comment by mesheets** [comment]
@solsjo, I concur with what your are seeing regarding `symtable + 4` and updated the `malloc()` as you suggested.  Valgrind now appears to pass with emu but now fails with an https://github.com/BrickBot/brickEmu/blob/25b76a5e9d0fd02c19d3bf1cebcfc7ec15f95770/ir-server.c#L77 message.

That had been a `printf()` call without a formatting string, so I changed it to a `puts()` (as I've seen usage of `printf()` without a formatting string flagged in other circumstances), but—unsurprisingly—that didn't make any difference.

While all of this was going on, another buffer overflow issue was [reported](https://github.com/BrickBot/brickEmu/issues/7#issue-2854668348) with emu in [sound_alsa.c](https://github.com/BrickBot/brickEmu/blob/master/sound_alsa.c) with Ubuntu 22.04 under WSL.  At first glance, I thought this might be something more specific to WSL, as the GitHub CI action running on Ubuntu 22.04 was passing.  In looking closer at the workflow file, I thought perhaps the libasound2-dev package might need to be explicitly added, but even after explicitly adding it, emu still passes the Valgrind action.

If `/usr/include/alsa` is not found, the `Makefile` sets up the build differently, and if alsa is removed, the buffer overflow error is reported to go away.

Here is the message that was reported while running Ubuntu 22.04 under WSL:
```
Starting: ./emu 45001
serverport=45001
Server started.
Starting server...Connected to server via 3.
Debugging-Server started on port 39829.
Connecting to Environment-Server at localhost port 45001.
BrickEmu: Preparing to Initialize Sound
Starting GUI...
Access type not available for playback: Invalid argument
BrickEmu: Sound Initialized
BrickEmu: Buttons Initialized
BrickEmu: LCD Initialized
BrickEmu: Initialization Complete
Overrun: 25
commit res: -77
Overrun: 54
commit res: -77
Overrun: 3
commit res: -77
Overrun: 196
==873==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffe3967583c at pc 0x56139a60e8b4 bp 0x7ffe39675770 sp 0x7ffe39675760
READ of size 4 at 0x7ffe3967583c thread T0
    # 0 0x56139a60e8b3 in sound_update /home/aconklin/brickEmu-master/sound_alsa.c:110
    # 1 0x56139a5f52c3 in t8_incr_tcnt /home/aconklin/brickEmu-master/timer8.c:180
    # 2 0x56139a5f54d3 in t8_update_time /home/aconklin/brickEmu-master/timer8.c:247
    # 3 0x56139a5f0c96 in wait_peripherals /home/aconklin/brickEmu-master/peripherals.c:282
    # 4 0x56139a5f0c96 in check_irq /home/aconklin/brickEmu-master/peripherals.c:327
    # 5 0x56139a5e814c in run_cpu /home/aconklin/brickEmu-master/h8300.c:292
    # 6 0x56139a5e736b in main /home/aconklin/brickEmu-master/main.c:109
    # 7 0x7f5623397d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f)
    # 8 0x7f5623397e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f)
    # 9 0x56139a5e7904 in _start (/home/aconklin/brickEmu-master/emu+0xc904)
Address 0x7ffe3967583c is located in stack of thread T0 at offset 108 in frame
    #0 0x56139a60e0ff in sound_update /home/aconklin/brickEmu-master/sound_alsa.c:40
   This frame has 6 object(s):
    [48, 56) 'mmap_areas' (line 55)
    [80, 88) 'offset' (line 56)
    [112, 120) 'frames' (line 56) <== Memory access at offset 108 underflows this variable
    [144, 152) 'delay' (line 84)
    [176, 180) 'step' (line 100)
    [192, 200) 'sptr' (line 99)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/aconklin/brickEmu-master/sound_alsa.c:110 in sound_update
Shadow bytes around the buggy address:
  0x1000472c6ab0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000472c6ac0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000472c6ad0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000472c6ae0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000472c6af0: 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1 f1 f1
=>0x1000472c6b00: 00 f2 f2 f2 00 f2 f2[f2]00 f2 f2 f2 00 f2 f2 f2
  0x1000472c6b10: 04 f2 00 f3 f3 f3 00 00 00 00 00 00 00 00 00 00
  0x1000472c6b20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000472c6b30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000472c6b40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000472c6b50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==873==ABORTING
Stream closed!
```

**Comment by solsjo** [comment]
Ack, I'm not entirely sure how the valgrind github actions work. E.g. do they run the binaries without flags? Or if they perform some static analysis.

If they only run valgrind on the binary without expected flags. Could it be that the github actions trigger behavior that would not be supported anyway?

Looking at the error message in this post, it seems like we read (or modify??) our  own stack variables unintentionally at sound_alsa.c:110?
As you say then perhaps the code makes assumptions that some calls would work, but they don't, if sound alsa is part of the build or not?

**Comment by solsjo** [comment]
It would seem that a set of these errors complain about mallocs that doesn't have a corresponding free? 
E.g.

'''
Error: ==10== 4,096 bytes in 1 blocks are still reachable in loss record 1 of 2
'''

**Comment by solsjo** [comment]
At this point I have a suggestion. It's more for the sake of book keeping.

I suggest that we remove the valgrind github actions from master.
Then we can see if master will execute the test and pass the pipeline.
Then we create new issue(s) were we add the github valgrind actions in pull requests that can be worked on until done.

This way it can be easier to separate, mostly working master, from special-case failures.
So that we can have an easier time knowing why master is failing.

**Comment by mesheets** [comment]
I've reconfigured the Valgrind actions to be variable based.  Right now, it defaults to false on push and pull_request events but allows setting this value to 'true' if manually kicking off the workflow.

Without Valgrind running, the workflow succeeds on all Ubuntu versions (20.04, 22.04, and 24.04).

**Comment by solsjo** [comment]
Ack! Great @mesheets
